manuals/vshadowinfo.1
.Dd June 26, 2026
.Dt VSHADOWINFO 1
.Os
.Sh NAME
.Nm vshadowinfo
.Nd determines information about a Volume Shadow Snapshot (VSS) volume
.Sh SYNOPSIS
.Nm vshadowinfo
.Op Fl o Ar offset
.Op Fl ahvV
.Ar source
.Sh DESCRIPTION
.Nm vshadowinfo
is a utility to determine information about a Volume Shadow Snapshot (VSS) \
volume
.Pp
.Nm vshadowinfo
is part of the
.Nm libvshadow
package.
.Nm libvshadow
is a library to access the Volume Shadow Snapshot (VSS) format
.Pp
.Ar source
is the source volume.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl a
shows allocation information
.It Fl h
shows this help
.It Fl o Ar offset
specify the volume offset in bytes
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# vshadowinfo /dev/sda1
vshadowinfo 20260626
.sp
Volume Shadow Snapshot information:
	Number of stores:	2
.sp
Store: 1
	Identifier		: 93db9c47-bb19-4004-836d-c3c835550b9a
	Shadow copy set ID	: 8bc68d0b-9df4-49e0-be4b-725dceaaefc8
	Creation time		: May 19, 2012 14:20:35.765721000 UTC
	Shadow copy ID		: 04057e11-d2d5-4d9c-8914-aeba832e467b
	Volume size		: 30002905088 bytes
	Attribute flags		: 0x00420009
.sp
Store: 2
	Identifier		: 93db9c47-bb19-4004-836d-c3c835550b9a
	Shadow copy set ID	: 90e19848-7436-4309-8899-4bbecf05a566
	Creation time		: May 19, 2012 16:13:58.750469800 UTC
	Shadow copy ID		: 6db2bc5e-9b95-4ed1-a493-091ee7d2e5e6
	Volume size		: 30002905088 bytes
	Attribute flags		: 0x00420009
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr vshadowmount 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libvshadow/issues
.Sh COPYRIGHT
Copyright (C) 2011-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
* rename libvshadow_volume_get_store to libvshadow_volume_get_store_by_index
* fix issue with multi threaded read test
  - store descriptor initialization?
* add error tolerance for missing NTFS backup volume header
* remove hard coded 0x4000 block size
  - use or pass io handle block size instead
* add tests
  - create test image
  - create test plan
  - test and add ignore overlay when reading from next store
* multi-threading support
  - what about abort value?
* beta
  - fix open issues
* add libfdata and libfcache for caching ?

API:
* allow access to block list
  - finish and expose libvshadow_store_get_block()
  - split block list and run-time btree
* get operating machine string
* get service machine string
* get size
* is open
* finish store close function

Error tollerance:
* add support for catalog entry 3 defined before 2

Format:
* how are multiple copies per set stored?
* the store file can be fragmented
  - block range list defines the fragments ?
* what about trailing volume data that windows is hiding 0x1000 bytes in size ?
  also seen in BDE, so likely something of NTFS

Tools:
* vshadowinfo
  - print attribute flags description
* vshadowmount
  - dokan fix volume size hints

20130503
* see `git log' for more recent change log
* added store read from file IO handle function
* vshadowmount small changes

20130501
* worked on vshadowmount Dokan support

20130417
* bug fix in dependencies

20130416
* code clean up

20130413
* updated dependencies
* updates and bug fixes in pyvshadow

20130412
* updated README files
* updated dependencies
* added tests directory

20130331
* updated dependencies
* code clean up
* pyvshadow updates
* vshadowtools now detect if there is a VSS signature first and bail out with
  a different error if not

20130304
* added PackageMaker files

20130217
* updated include/types.h

20130205
* fixed typo in vhsadowmount

20130131
* worked on pyvshadow

20130130
* worked on exposing block descriptors via vshadowinfo
* worked on pyvshadow

20130129
* worked on exposing block descriptors via API
* worked on pyvshadow

20130128
* worked on exposing block descriptors via API

20130127
* worked on exposing block descriptors via API
* removed LIBVSHADOW_STORE_FLAG_IO_HANDLE_MANAGED flags

20130125
* worked on exposing block descriptors via API

20130113
* 2013 update
* updated dependencies
* updated msvscpp files
* added vshadowmount.1 man page

20121122
* pyvshadow: code clean up

20121118
* pyvshadow: switched to memory_allocate and memory_free to
  PyMem_Malloc and PyMem_Free

20121107
* updates msvscpp 2010 build

20121104
* pyvshadow: fixes for 32-bit build

20121103
* fix in spec file
* updated dependencies

20121102
* pyvshadow: fix for Mac OS X build
* updated msvscpp files

20121017
* code clean up

20121016
* pyvshadow: bug fixes
  - Missing Py_None increment reference
  - added increment/decrement reference of volume object in store

20121012
* pyvshadow: added creation time as integer function

20121011
* minor changes

20121010
* made get store more restrictive

20121009
* added store get size function for python binding

20121008
* updated dpkg and spec files

20121007
* added store get offset function
* worked on Python bindings
* fix for dpkg files docs

20121006
* worked on Python bindings

20120922
* small changes

20120921
* improved dealing with NTFS backup header for volume size detection

20120915
* improved catalog entry handling
* changes for libcdata
* worked on man pages
* worked on removing hard coded block size
* code clean up
* worked on support for catalog size corruption issue

20120914
* improved catalog entry handling
* added -X option to vshadowmount to pass options to fuse layer

20120912
* vshadowmount: replaced EBADFD by EIO for OSX

20120911
* fixed bug regarding missing block size correction

20120905
* updated libbfio

20120904
* code clean up

20120903
* changed overlay handling
* code clean up

20120902
* replaced data types by libcdata

20120901
* split off block list and block tree usage of block descriptors
  - block tree can override info in block descriptors while block list should not
* moved overlays from array to single block descriptor
* removed metadata offset list

20120831
* fixes and improvements in block tree
* changed initial bitmap to current bitmap
* code clean up

20120830
* fixes and improvements in btree

20120829
* fixes in btree
* added more debug info

20120828
* added more debug info

20120826
* code clean up
* initial version of msvscpp files

20120825
* improvement to handling not in-use blocks

20120824
* added support to handle duplicate reverse block tree blocks
* volume size is now determined from store descriptors in catalog
* code clean up

20120823
* worked on multi snapshot handling

20120822
* worked on block descriptors
* code clean up

20120821
* worked on block descriptors

20120820
* applied updates
* code clean

20120721
* added libcsplit
* applied updates
* switched library order in tools

20120603
* worked on block behavior

20120528
* worked on block behavior
* bug fix in mount handle

20120527
* worked on block behavior

20120526
* worked on block behavior

20120525
* worked on block behavior

20120524
* worked on block behavior
* re-introduced block wrap

20120523
* worked on block behavior

20120522
* worked on block behavior

20120520
* worked on block behavior

20120519
* worked on initial and previous bitmap

20120517
* worked on initial bitmap
* optimized reading bitmap
* fixed memory leaks
* worked on windows 8 support

20120515
* added support for volume without shadows

20120513
* fixed mingw build - linking of libbfio

20120511
* updated dependencies
* added fuse destroy function

20120415
* updated dependencies
* added debug of attribute flags
* changed debug info to print less zero byte data

20120225
* 2012 update

20111218
* updated libcstring, libsystem

20111213
* worked on initial bitmap
* worked on block descriptor handling

20111212

vshadowtools/vshadowinfo.c
/*
 * Shows information obtained from a Volume Shadow Snapshot (VSS) volume.
 *
 * Copyright (C) 2011-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <system_string.h>
#include <types.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "info_handle.h"
#include "vshadowtools_getopt.h"
#include "vshadowtools_libcerror.h"
#include "vshadowtools_libclocale.h"
#include "vshadowtools_libcnotify.h"
#include "vshadowtools_libvshadow.h"
#include "vshadowtools_output.h"
#include "vshadowtools_signal.h"
#include "vshadowtools_unused.h"

info_handle_t *vshadowinfo_info_handle = NULL;
int vshadowinfo_abort                  = 0;

/* Signal handler for vshadowinfo
 */
void vshadowinfo_signal_handler(
      vshadowtools_signal_t signal VSHADOWTOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "vshadowinfo_signal_handler";

	VSHADOWTOOLS_UNREFERENCED_PARAMETER( signal )

	vshadowinfo_abort = 1;

	if( vshadowinfo_info_handle != NULL )
	{
		if( info_handle_signal_abort(
		     vshadowinfo_info_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal info handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use vshadowinfo to determine information about a Volume Shadow Snapshot (VSS) volume.";

	vshadowtools_option_t options[ ] = {
		{ 'a', NULL, "shows allocation information" },
		{ 'h', NULL, "shows this help" },
		{ 'o', "offset", "specify the volume offset in bytes" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source volume" },
	};
	system_character_t options_string[ 32 ];

	libvshadow_error_t *error           = NULL;
	system_character_t *option_offset   = NULL;
	system_character_t *source          = NULL;
	char *program                       = "vshadowinfo";
	system_integer_t option             = 0;
	uint8_t show_allocation_information = 0;
	int number_of_options               = (int) ( sizeof( options ) / sizeof( vshadowtools_option_t ) );
	int result                          = 0;
	int verbose                         = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "vshadowtools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( vshadowtools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	vshadowtools_output_version_fprint(
	 stdout,
	 program );

	if( vshadowtools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = vshadowtools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				vshadowtools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'a':
				show_allocation_information = 1;

				break;

			case (system_integer_t) 'h':
				vshadowtools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'o':
				option_offset = optarg;

				break;

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				vshadowtools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source volume.\n" );

		vshadowtools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	libcnotify_verbose_set(
	 verbose );
	libvshadow_notify_set_stream(
	 stderr,
	 NULL );
	libvshadow_notify_set_verbose(
	 verbose );

	if( info_handle_initialize(
	     &vshadowinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize info handle.\n" );

		goto on_error;
	}
#if defined( __clang_analyzer__ )
	__builtin_assume( vshadowinfo_info_handle != NULL );
#endif
	if( option_offset != NULL )
	{
		if( info_handle_set_volume_offset(
		     vshadowinfo_info_handle,
		     option_offset,
		     &error ) != 1 )
		{
			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );

			fprintf(
			 stderr,
			 "Unsupported volume offset defaulting to: %" PRIi64 ".\n",
			 vshadowinfo_info_handle->volume_offset );
		}
	}
	vshadowinfo_info_handle->show_allocation_information = show_allocation_information;

	result = info_handle_open_input(
	          vshadowinfo_info_handle,
	          source,
	          &error );

	if( result == -1 )
	{
		fprintf(
		 stderr,
		 "Unable to open: %" PRIs_SYSTEM ".\n",
		 source );

		goto on_error;
	}
	else if( result == 0 )
	{
		fprintf(
		 stderr,
		 "No Volume Shadow Snapshots found.\n" );

		goto on_error;
	}
	if( info_handle_volume_fprint(
	     vshadowinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to print volume information.\n" );

		goto on_error;
	}
	if( info_handle_close(
	     vshadowinfo_info_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close info handle.\n" );

		goto on_error;
	}
	if( info_handle_free(
	     &vshadowinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free info handle.\n" );

		goto on_error;
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( vshadowinfo_info_handle != NULL )
	{
		info_handle_free(
		 &vshadowinfo_info_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

